### PR TITLE
[repo-tools] Command to create/update `catalog-info.yaml` files for Backstage Packages

### PR DESCRIPTION
**Pull request by iamEAP**
## tl;dr
A new `backstage-repo-tools` command to help maintain `catalog-info.yaml` files representing Backstage plugin packages, including a standard for such components.

The command can be invoked very simply:

```sh
# See what would happen
backstage-repo-tools generate-catalog-info --dry-run

# Actually do the thing
backstage-repo-tools generate-catalog-info
```

## What / Why

This will create any missing `catalog-info.yaml` files, and update existing `catalog-info.yaml` files so that they conform to a proposed minimum standard:

```yaml
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: backstage-plugin-foo # must always be slugified version of "name" key in package.json
  title: '@backstage/plugin-foo' # must always be the "name" key in package.json
spec:
  type: backstage-frontend-plugin # must always be `backstage-` followed by the "backstage.role" key in package.json
  owner: maintainers # should match a CODEOWNER who owns the file/dir, stripped of any preceding org
```

Any extra metadata beyond the above minimum set of metadata will be maintained when the command is re-run (e.g. `spec.lifecycle`, annotations, etc).

Although this PR does not add generated `catalog-info.yaml` files _yet_, I intend to add them in a follow-up PR, alongside a pre-commit hook and build step that ensures that they remain up-to-date with respect to `package.json` and `CODEOWNER` changes, where the goal would be to experiment with their utility in the monorepo.

This is an update to (and replacement for) #17934

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by aitherios — APPROVED** [review]

**Review by adamdmharvey — APPROVED** [review]

**Comment by iamEAP** [comment]
Thanks for the call-out @awanlin!  After reviews are in order and before this gets merged in, I'll be sure to do one last rebase and re-run of the command (updated for now, anyway)

**Comment by iamEAP** [comment]
@benjdlambert given you're assigned to this: wondering if rather than getting all reviews in for this from everyone who owns a thing, maybe it makes more sense for you to look over the `repo-tools` code and merge it in without explicit approvals?  Once it's in, folks can self-update anyway.

File: packages/app-defaults/catalog-info.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+apiVersion: backstage.io/v1alpha1
+kind: Component
+metadata:
+  name: backstage-app-defaults
+  title: '@backstage/app-defaults'
+  description: Provides the default wiring of a Backstage App
+spec:
+  lifecycle: experimental
+  type: backstage-package
+  backstage_role: web-library
```
Comment (pjungermann) [inline packages/app-defaults/catalog-info.yaml:10]: I'm wondering about:

1. type `backstage-package` vs `library` as in the default model docs. The extension field `backstage_role` could be enough indication. 

2. `backstage_role` seems not to follow the camelCase naming pattern. I would go for ` backstageRole` for consistency reasons.

**Review by pjungermann — COMMENTED** [review]
Thank you for your work, @iamEAP !

**Review by kuangp — APPROVED** [review]

**Comment by vinzscam** [comment]
I understand the purpose of the new command in case people wants to use this approach in their Backstage instance, but I've some troubles understanding the purpose of shipping a `catalog-info.yaml` for each plugin contained in the Backstage monorepo. Wasn't enough to just ship the script of the new command?

File: packages/app-defaults/catalog-info.yaml (hunk 1 above)
Comment (iamEAP) [reply packages/app-defaults/catalog-info.yaml:10]: Great call-out re: `backstage_role`!  I'll update.

On `backstage-package` vs. `library`: we definitely considered this.

Our hesitation with `library` is that, although open source plugins resemble what are typically modeled as library components (e.g. they are packaged code, distributed to some kind of registry, with versioned releases), innersource plugins do not map well to this paradigm (they are versionless pieces of code built and deployed directly alongside an existing service--an organization's Backstage instance).  Maybe less important, but usage patterns of libraries are also different from these backstage packages: libraries you might expect to see used across a variety of systems in a variety of contexts, while Backstage packages are almost exclusively used within Backstage (and often only used by a particular plugin).

Hope is that this will simplify entity pages for these components (just leverage `isComponentType()` and let it sit alongside other types without having to break down a particular type further), and that it might simplify targeting these types of components for the purposes of evaluating tech health.

**Comment by iamEAP** [comment]
> I've some troubles understanding the purpose of shipping a `catalog-info.yaml` for each plugin contained in the Backstage monorepo. Wasn't enough to just ship the script of the new command and then adopters can invoke the script to generate `catalog-info.yaml` for their plugins?

Thanks for the question @vinzscam.  I guess my response kind of depends on where your primary concerns are coming from, but I can give at least an initial thought:

The aspiration with the command is that this becomes the sort of de-facto way to track backstage packages in the software catalog.  A related philosophy is that every logical piece of software should be owned.  So the idea is to stamp out `catalog-info.yaml` files for every logical package up-front, and ensure that the ownership remains in sync with the actual code ownership going forward.

**Comment by vinzscam** [comment]
> > I've some troubles understanding the purpose of shipping a `catalog-info.yaml` for each plugin contained in the Backstage monorepo. Wasn't enough to just ship the script of the new command and then adopters can invoke the script to generate `catalog-info.yaml` for their plugins?
> 
> Thanks for the question @vinzscam. I guess my response kind of depends on where your primary concerns are coming from, but I can give at least an initial thought:
> 
> The aspiration with the command is that this becomes the sort of de-facto way to track backstage packages in the software catalog. A related philosophy is that every logical piece of software should be owned. So the idea is to stamp out `catalog-info.yaml` files for every logical package up-front, and ensure that the ownership remains in sync with the actual code ownership going forward.

thanks for the context.
I like the idea of tracking the ownership of every single plugin in a Backstage app. But in the case of this monorepo, these are plugins that are pulled in adopters' Backstage instances as kind of dependencies, so they shouldn't be tracked. Usually you don't track Maven dependencies or package.json dependencies used by each service.

**Comment by iamEAP** [comment]
> But in the case of this monorepo, these are plugins that are pulled in adopters' Backstage instances as kind of dependencies, so they shouldn't be tracked. Usually you don't track Maven dependencies or package.json dependencies used by each service.

:100: agree!   The intention is not for everyone in the world to track every plugin in _this_ monorepo (though you likely would in your internal Backstage monorepo).  Rather, you would configure whatever discovery ingestion mechanism to only pick out the packages you actually feel you own/maintain.

Additional work would likely be required anyway (e.g. a custom catalog processor) to map external GitHub org/team/individual names to internal group/user names.

**Comment by iamEAP** [comment]
> _which might introduce the risk to regenerate all catalog-info.yaml again_

The hope is that what's proposed here will more or less be all the metadata that's ever tracked in this particular repo; the risk of major changes needed seems quite low (package names won't change, the type won't change, the role won't change.  The only thing likely to change is ownership, and the hope is that [the hook](https://github.com/backstage/backstage/pull/18354/commits/f64958c2523403c0d30469f768b45033979c0701) will make that basically transparent.

**Comment by awanlin** [comment]
@iamEAP, could we add annotations to these? For example the Azure DevOps and Linguist plugins have their own annotations and it would be nice to have them setup in their own `catalog-info.yaml` files that way it would be easy to test things in this repo or taking it further pulling these into the Demo site to be able to who the feature of the plugins.

**Comment by benjdlambert** [comment]
Hey @iamEAP! Gonna whack the `after-vacations` label on this because I want to discuss the schema a little bit with other maintainers and it feels like there's still some ongoing conversion that could do with resolving first before we get this in if that's OK? :prayt:

**Comment by freben** [comment]
I added this to the [Catalog SIG agenda](https://docs.google.com/document/d/1xg57CDfX2R5jTGbbWANzifm2szEb664kuTwfEEvKgZU/edit) to talk more about how to best proceed.

File: packages/repo-tools/src/commands/generate-catalog-info/generate-catalog-info.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  const badDesc = yamlJson.metadata.description !== packageJson.description;
+
+  if (badOwner || badTitle || badName || badType || badDesc) {
+    const owner = badOwner
+      ? getOwnerFromCodeowners(codeowners, yamlPath)
+      : yamlJson.spec?.owner;
+    const newJson = createOrMergeEntity(packageJson, owner, yamlJson);
+    return dryRun
+      ? Promise.resolve(console.error(`Update ${relativePath('.', yamlPath)}`))
+      : writeFile(yamlPath, yamlOverwrite(yamlString, newJson));
```
Comment (Rugvip) [inline packages/repo-tools/src/commands/generate-catalog-info/generate-catalog-info.ts:161]: This thing is interesting 😁 curious how well it handles comments etc, gonna have to play around with it a bit

File: packages/repo-tools/src/commands/generate-catalog-info/generate-catalog-info.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+              codeowners,
+              dryRun,
+            });
+            return;
+          }
+
+          throw e;
+        }
+
+        fixCatalogInfoYaml({
```
Comment (Rugvip) [inline packages/repo-tools/src/commands/generate-catalog-info/generate-catalog-info.ts:74]:
```suggestion
        await fixCatalogInfoYaml({
```

Right? Throwing this in to potentially ship asap :p

File: packages/repo-tools/src/commands/generate-catalog-info/generate-catalog-info.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    const owner = badOwner
+      ? getOwnerFromCodeowners(codeowners, yamlPath)
+      : yamlJson.spec?.owner;
+    const newJson = createOrMergeEntity(packageJson, owner, yamlJson);
+    return dryRun
+      ? Promise.resolve(console.error(`Update ${relativePath('.', yamlPath)}`))
+      : writeFile(yamlPath, yamlOverwrite(yamlString, newJson));
+  }
+
+  return Promise.resolve();
```
Comment (Rugvip) [inline packages/repo-tools/src/commands/generate-catalog-info/generate-catalog-info.ts:164]: I find it nicer to mark functions as `async` instead 😅

**Review by Rugvip — APPROVED** [review]
Nice! 🎉 

Let's give it a spin 😁 :shipit:
